beluga/include/beluga/algorithm/distance_map.hpp

```diff
@@ -92,6 +93,56 @@ auto nearest_obstacle_distance_map(
   return distance_map;
 }
 
+/// A distance map function considering unknown values.
+template <class Range, class DistanceFunction, class NeighborsFunction>
+auto nearest_obstacle_unknown_distance_map(
+    Range&& obstacle_map,
+    DistanceFunction&& distance_function,
+    NeighborsFunction&& neighbors_function,
+    const double unknown_distance_map) {
+  struct IndexPair {
+    std::size_t nearest_obstacle_index;
+    std::size_t index;
+  };
+
+  using DistanceType = std::invoke_result_t<DistanceFunction, std::size_t, std::size_t>;
+  auto distance_map = std::vector<DistanceType>(ranges::size(obstacle_map));
+  auto visited = std::vector<bool>(ranges::size(obstacle_map), false);
+
+  auto compare = [&distance_map](const IndexPair& first, const IndexPair& second) {
+    return distance_map[first.index] > distance_map[second.index];
+  };
+  auto queue = std::priority_queue<IndexPair, std::vector<IndexPair>, decltype(compare)>{compare};
+
+  for (auto [index, enum_data] : ranges::views::enumerate(obstacle_map)) {
+    if (enum_data == beluga::GridStatus::kOccupied) {
+      visited[index] = true;
+      distance_map[index] = 0;
+      queue.push(IndexPair{index, index});  // The nearest obstacle is itself
+    }
+  }
+
+  while (!queue.empty()) {
+    auto parent = queue.top();
+    queue.pop();
+    for (const std::size_t index : neighbors_function(parent.index)) {
+      if (!visited[index]) {
+        visited[index] = true;
+        distance_map[index] = distance_function(parent.nearest_obstacle_index, index);
+        queue.push(IndexPair{parent.nearest_obstacle_index, index});
+      }
+    }
+  }
+
+  for (auto [index, enum_data] : ranges::views::enumerate(obstacle_map)) {
```

Comment:
@DPR00 meta: can't we decouple this so that we reuse `nearest_obstacle_distance_map`? If distance map matches 1-to-1 with grid data, why not zip it with grid ternary value and transform it downstream?

Reply:
This made me realize that we may not need `ternary_data()`. PR updated.